backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/query/impl/VectorSimilarityFilterQuery.java

```diff
@@ -0,0 +1,222 @@
+		// we've already converted distance/similarity to a score, but now if the underlying query is boosting the score,
+		// we'd want to boost our converted one as well to get the expected matches:
+		return new SimilarityWeight( query.createWeight( searcher, scoreMode, boost ), similarityAsScore * boost );
+	}
+
+	@Override
+	public void visit(QueryVisitor visitor) {
+		visitor.visitLeaf( this );
+	}
+
+	@Override
+	public String toString(String field) {
+		return getClass().getName() + "{" +
+				"query=" + query +
+				", similarityLimit=" + similarityAsScore +
+				'}';
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if ( this == o ) {
+			return true;
+		}
+		if ( o == null || getClass() != o.getClass() ) {
+			return false;
+		}
+		VectorSimilarityFilterQuery that = (VectorSimilarityFilterQuery) o;
+		return Float.compare( similarityAsScore, that.similarityAsScore ) == 0 && Objects.equals( query, that.query );
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash( query, similarityAsScore );
+	}
+
+	private static float floatSimilarityDistanceToScore(float distance, VectorSimilarityFunction similarityFunction) {
+		switch ( similarityFunction ) {
+			case EUCLIDEAN:
+				return 1.0f / ( 1.0f + distance * distance );
+			case DOT_PRODUCT:
+			case COSINE:
+				return ( 1.0f + distance ) / 2.0f;
+			case MAXIMUM_INNER_PRODUCT:
+				return VectorUtil.scaleMaxInnerProductScore( distance );
+			default:
+				throw new AssertionFailure( "Unknown similarity function: " + similarityFunction );
+		}
+	}
+
+	private static float byteSimilarityDistanceToScore(float distance, int dimension,
+			VectorSimilarityFunction similarityFunction) {
+		switch ( similarityFunction ) {
+			case EUCLIDEAN:
+				return 1.0f / ( 1.0f + distance * distance );
+			case DOT_PRODUCT:
+				return 0.5f + distance / (float) ( dimension * ( 1 << 15 ) );
+			case COSINE:
+				return ( 1.0f + distance ) / 2.0f;
+			case MAXIMUM_INNER_PRODUCT:
+				return VectorUtil.scaleMaxInnerProductScore( distance );
+			default:
+				throw new AssertionFailure( "Unknown similarity function: " + similarityFunction );
+		}
+	}
+
+	private static class SimilarityWeight extends FilterWeight {
+		private final float similarityAsScore;
+
+		protected SimilarityWeight(Weight weight, float similarityAsScore) {
+			super( weight );
+			this.similarityAsScore = similarityAsScore;
+		}
+
+		@Override
+		public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+			Explanation explanation = super.explain( context, doc );
+			if ( explanation.isMatch() && similarityAsScore > explanation.getValue().floatValue() ) {
+				return Explanation.noMatch( "Similarity limit is greater than the vector similarity.", explanation );
+			}
+			return explanation;
+		}
+
+		@Override
+		public Scorer scorer(LeafReaderContext context) throws IOException {
+			Scorer scorer = super.scorer( context );
+			if ( scorer == null ) {
+				return null;
```

Comment:
I was running some tests locally and noticed that sometimes the super-scorer can be null, leading to a `NullPointerExeception` in `MinScoreScorer`. So I've added this if statement. See the knn scorer part that would return a `null`

https://github.com/apache/lucene/blob/2a0b7f2056745a2cafbda8cafe47eb7bcb94d28a/lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java#L349-L351